### PR TITLE
[FAD-5924] Relative date re-calculation on report load/refresh

### PR DESCRIPTION
**Pull request by jasonrhodes**
Adjust the summary report dates until you get a custom range (use last hour for easiest test), then wait a few minutes, refresh the page. Times will update to _current_, re-calculated last hour.

Click share modal when you're on a relative range and you'll see a prompt to pin the range (which only changes the range={relativeRange} to range=custom and back. 

When on a custom range, the share modal shows no note or lock button.

Also, the share button is now disabled while the chart is reloading (prevents weird in-between modal state)

### CODE REVIEW

**Comment by jasonrhodes** [comment]
First draft here is seen below, going to experiment with a few other UX ideas soon.
share-modal-relative-toggle

File: src/pages/reports/components/ShareModal.js [hunk 1]
```diff
@@ -1,17 +1,42 @@
...
+
+    return (
+      <Modal open={open}>
+        {open && <WindowEvent event='keydown' handler={handleToggle} />}
+        <Panel title='Share this report'>
+          <Panel.Section>
+            {isRelative && !pinned && <p>Note: This link will always load this report with a relative range ({relativeRange.label}).</p>}
+            {isRelative && pinned && <p>Note: This link will always load this report with specific dates ({from} to {to}).</p>}
+            <CopyField
+              value={`${window.location.href}?${search}`}
```
Comment (jonambas) [inline src/pages/reports/components/ShareModal.js]: Getting double search here. Think `window.location.href` includes everything.

`http://localhost:3100/reports/summary?from=2017-12-05T20:38:47Z&metrics=count_targeted&metrics=count_rendered&metrics=count_accepted&metrics=count_bounce&range=30days&to=2018-01-04T20:38:47Z?from=2017-12-05T20:38:47Z&metrics=count_targeted&metrics=count_rendered&metrics=count_accepted&metrics=count_bounce&range=30days&to=2018-01-04T20:38:47Z`

File: src/helpers/reports.js [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
  * Creates search options object from shared report options. Page specific options not included (ie. summary chart selected metrics)
  * @param  {Object} filters - reportFilters state
  * @return {Object} - formatted search options object
  */
-function getFilterSearchOptions(filters) {
+export function getFilterSearchOptions(filters) {
   return {
     from: moment(filters.from).utc().format(),
     to: moment(filters.to).utc().format(),
+    range: filters.relativeRange,
```
Comment (jonambas) [inline src/helpers/reports.js]: Would it make sense to only include to/from if set to `range` is `custom`, and leave them out if a `range` is set? It would make the share url look a little cleaner

File: src/pages/reports/summary/SummaryPage.js [hunk 3]
```diff
@@ -70,24 +73,22 @@ class SummaryReportPage extends Component {
...
       metrics: chart.metrics.map((metric) => metric.key),
       ...getFilterSearchOptions(filters)
     };
 
-    const { link, search } = getShareLink(options);
-
-    this.setState({ link });
-    history.replace({ pathname: '/reports/summary', search });
+    this.setState({ query });
+    history.replace({ pathname: '/reports/summary', search: qs.stringify(query, { encode: false }) });
```
Comment (jonambas) [inline src/pages/reports/summary/SummaryPage.js]: Will need to remove `getShareLink` and do this on every other report page as well.

File: src/pages/reports/components/ShareModal.js [hunk 4]
```diff
@@ -1,17 +1,42 @@
...
+    const { open, query, handleToggle } = this.props;
+    const relativeRange = relativeDateOptions.find((item) => item.value === query.range);
+    const isRelative = relativeRange && query.range !== 'custom';
+    const { pinned } = this.state;
+    const pinnedRange = pinned ? { range: 'custom' } : {};
+
+    const search = qs.stringify({ ...query, ...pinnedRange }, { encode: false });
+    const connectLeft = !isRelative ? null : <Button primary={this.state.pinned} onClick={() => this.setState({ pinned: !pinned })}><Icon name='Lock' /></Button>;
+    const from = moment(query.from).format('d MMM Y h:mma');
+    const to = moment(query.to).format('d MMM Y h:mma');
```
Comment (jonambas) [inline src/pages/reports/components/ShareModal.js]: I know you're not done here, but this is a little hard to follow

File: src/helpers/reports.js (hunk 2 above)
Comment (jasonrhodes) [reply src/helpers/reports.js]: I removed them from the link inside the share modal when it's relative and not pinned.

**Comment by jasonrhodes** [comment]
screen shot 2018-01-05 at 5 28 44 pm

Simplified a bit for now.

File: src/helpers/reports.js [hunk 5]
```diff
@@ -27,7 +29,7 @@ function getFilterSearchOptions(filters) {
  *     link - full url
  *   }
  */
-function getShareLink(options) {
+export function getShareLink(options) {
```
Comment (jasonrhodes) [inline src/helpers/reports.js]: Can remove

**Comment by jonambas** [comment]
I saw @bizob2828 refactored some of the same code a little: https://github.com/SparkPost/2web2ui/pull/231/files#diff-f3da886548ee5040c8ac97df0ed2a055L31

File: src/pages/reports/components/ShareModal.js [hunk 6]
```diff
@@ -1,17 +1,78 @@
...
-        <Button primary onClick={handleToggle}>Done</Button>
-      </Panel.Section>
-    </Panel>
-  </Modal>
-);
-
-export default ShareModal;
+export default class ShareModal extends Component {
+  state = {
+    pinned: true
```
Comment (jonambas) [inline src/pages/reports/components/ShareModal.js]: I'm still wondering where this assumption that people by default want to share calculated dates. I guess maybe because this is how the old app works, but 🤔 

There are going to be a number of folks who skip the modal completely and just copy the url from the address bar (which defaults to the opposite of this modal if using relative range). Should we make sure the url and default share modal link are always the same? We would need to 1) setting this to false 2) hide to/from from before options get to history.replace

File: src/pages/reports/components/ShareModal.js [hunk 7]
```diff
@@ -1,17 +1,78 @@
...
+    const { pinned } = this.state;
+    const relativeRange = relativeDateOptions.find((item) => item.value === query.range);
+    const isRelative = relativeRange && query.range !== 'custom';
+
+    if (!isRelative) {
+      return null;
+    }
+
+    return (
+      <div style={{ marginBottom: '30px' }}>
```
Comment (jonambas) [inline src/pages/reports/components/ShareModal.js]: You can avoid this by wrapping the textfield & checkbox in a panel.section, and the done button in its own panel.section

File: src/pages/reports/components/ShareModal.js (hunk 6 above)
Comment (jasonrhodes) [reply src/pages/reports/components/ShareModal.js]: My assumption is that the most common use-case for sharing a report is so you can share data you are looking at, some spike or change etc etc. If you accidentally share that as relative, there's no way for the receiver of the link to know what dates you were looking at. 

By defaulting the share link to fixed dates, you have to *decide* to switch back to the relative version. If you accidentally share the fixed version, it's very easy for the receiver of the link to simply pick the relative range they'd rather see. 

In my opinion, if you interact with URL directly, you're on your own and I'm totally ok with that. I think that's actually the reason for the share modal, to provide sugar and better defaults than just copying the URL.

File: src/pages/reports/components/ShareModal.js (hunk 7 above)
Comment (jasonrhodes) [reply src/pages/reports/components/ShareModal.js]: Had just added this as a quick fix and forgot to swap out, will do thanks.

File: src/pages/reports/components/ShareModal.js (hunk 6 above)
Comment (jonambas) [reply src/pages/reports/components/ShareModal.js]: Hmm okay, makes sense 👍

File: src/pages/reports/bounce/BouncePage.js [hunk 8]
```diff
@@ -105,19 +100,24 @@ export class BouncePage extends Component {
...
+        <Filters
+          refresh={this.handleRefresh}
+          onShare={this.handleModalToggle}
+          shareDisabled={chartLoading}
+        />
         { this.renderChart() }
         <Panel title='Bounced Messages' className='ReasonsTable'>
           { this.renderCollection() }
         </Panel>
         <ShareModal
```
[the remark this answers is not in the document]
Comment (jasonrhodes) [reply src/pages/reports/bounce/BouncePage.js]: Good call, I'll take a look at that -- gonna file a ticket now related to the other report refactor/cleanup ticket and to this one.
